File: Sources/UsageWidget/Domain/TokenStats.swift
import Foundation

/// Optional secondary token/cost detail, shown in the collapsible "Details" section.
/// May be `nil` when a provider doesn't expose token counts through its live endpoint.
struct TokenStats: Hashable, Codable, Sendable {
    var inputTokens: Int?
    var outputTokens: Int?
    var cacheReadTokens: Int?
    var cacheWriteTokens: Int?
    var estimatedCostUSD: Double?
    var since: Date?

    init(
        inputTokens: Int? = nil,
        outputTokens: Int? = nil,
        cacheReadTokens: Int? = nil,
        cacheWriteTokens: Int? = nil,
        estimatedCostUSD: Double? = nil,
        since: Date? = nil
    ) {
        self.inputTokens = inputTokens
        self.outputTokens = outputTokens
        self.cacheReadTokens = cacheReadTokens
        self.cacheWriteTokens = cacheWriteTokens
        self.estimatedCostUSD = estimatedCostUSD
        self.since = since
    }

    /// Sum of all known token buckets, or `nil` if none are present.
    var totalTokens: Int? {
        let parts = [inputTokens, outputTokens, cacheReadTokens, cacheWriteTokens].compactMap { $0 }
        return parts.isEmpty ? nil : parts.reduce(0, +)
    }

    var hasAnyValue: Bool {
        totalTokens != nil || estimatedCostUSD != nil
    }
}

File: windows/src/Core/Domain/TokenStats.cs
namespace LLMUsageWidget.Core.Domain;

/// <summary>Optional secondary token/cost detail, shown in the collapsible "Details" section.
/// May be null when a provider doesn't expose token counts through its live endpoint.</summary>
public sealed record TokenStats(
    int? InputTokens = null,
    int? OutputTokens = null,
    int? CacheReadTokens = null,
    int? CacheWriteTokens = null,
    double? EstimatedCostUsd = null,
    DateTimeOffset? Since = null)
{
    /// <summary>Sum of all known token buckets, or null if none are present.</summary>
    public int? TotalTokens
    {
        get
        {
            int[] parts = new[] { InputTokens, OutputTokens, CacheReadTokens, CacheWriteTokens }
                .Where(p => p.HasValue).Select(p => p!.Value).ToArray();
            return parts.Length == 0 ? null : parts.Sum();
        }
    }

    public bool HasAnyValue => TotalTokens != null || EstimatedCostUsd != null;
}
